File: leetcode/python/121_best_stock_profit.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        curMin = prices[0]
        profit: int = 0
        for n in prices[1:]:
            profit = max(profit, n - curMin)
            curMin = min(n, curMin)
        return profit

s = Solution()
print(s.maxProfit([7,1,5,3,6,4]))
print(s.maxProfit([7,6,4,3,1]))
        

File: leetcode/c++/121_best_stock_profit.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int curMin = prices[0];
        int maxProfit = 0;
        for (auto price : prices) {
            if (price < curMin) {
                curMin = price;
            } else {
                maxProfit = max(maxProfit, price - curMin);
            }
        }
        return maxProfit;
    }
};